File: src/android/CDVBMSAnalytics.java
/*
    Copyright 2015 IBM Corp.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
        http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/
package com.ibm.mobilefirstplatform.clientsdk.cordovaplugins.core;

import com.ibm.mobilefirstplatform.clientsdk.android.core.api.Response;
import com.ibm.mobilefirstplatform.clientsdk.android.core.api.ResponseListener;
import com.ibm.mobilefirstplatform.clientsdk.android.analytics.api.*;
import com.ibm.mobilefirstplatform.clientsdk.android.logger.api.Logger;

import android.util.Log;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import android.app.Application;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import java.util.Arrays;

public class CDVBMSAnalytics extends CordovaPlugin {

    private static final Logger analyticsLogger = Logger.getLogger(Logger.INTERNAL_PREFIX + "CDVBMSAnalytics");
    private String errorEmptyArg = "Expected non-empty string argument.";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        analyticsLogger.debug("execute() : action = " + action);

        if("enable".equals(action)) {
            Analytics.enable();
            callbackContext.success();
            return true;
        } else if("disable".equals(action)) {
            Analytics.disable();
            callbackContext.success();
            return true;
        } else if("isEnabled".equals(action)) {
            //TODO: This does not exist for Android SDK
            callbackContext.error("Not yet implemented");
            return true;
        } else if("send".equals(action)) {
            this.send(callbackContext);
            return true;
        } else if("logEvent".equals(action)) {
            // TODO: Not yet implemented for the Android SDK
            callbackContext.error("Not yet implemented");
            return true;
        } else if("initialize".equals(action)){
            this.init(args, callbackContext);
            return true;
        } else if("log".equals(action)){
            this.log(args, callbackContext);
            return true;
        } else if("setUserIdentity".equals(action)){
            this.setUserIdentity(args, callbackContext);
            return true;
        }
        return false;
    }

    /**
     *
     * Specify current application user.  This value will be hashed to ensure privacy.
     * If your application does not have user context, then nothing will happen.
     *
     * @param args  JSONArray that contains the argument the to initialize Analytics
     * @param callbackContext
     */

    public void setUserIdentity(final JSONArray args, final CallbackContext callbackContext){
        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    String userIdentity = args.getString(0);
                    Analytics.setUserIdentity(userIdentity);
                } catch(JSONException e){
                    analyticsLogger.error("setUserIdentity :: Analytics failed to set user identity. " +
                            "Please review arguments");
                    callbackContext.error(e.getMessage());
                }

            }
        });
    }

    /**
     * Sends the Analytics logs to the server
     * @param callbackContext Callback that will indicate whether the request succeeded or failed
     */
    public void send(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                Analytics.send(new ResponseListener() {
                    @Override
                    public void onSuccess(Response response) {
                        callbackContext.success();
                    }

                    @Override
                    public void onFailure(Response failResponse, Throwable t, JSONObject extendedInfo) {
                        callbackContext.error(failResponse.toString());
                    }

                });
            }
        });


    }

    /**
     * Initialize BMSAnalytics API
     * @param args  JSONArray that contains the argument the to initialize Analytics
     * @param callbackContext
     */
    public void init(final JSONArray args, final CallbackContext callbackContext){

        final Application app = cordova.getActivity().getApplication();
        try {
            String applicationName = args.getString(0);
            String clientApiKey = args.getString(1);
            boolean hasContext = args.getBoolean(2);
            JSONArray deviceEventsArray = args.getJSONArray(3);
            Analytics.DeviceEvent[] devices = new Analytics.DeviceEvent[deviceEventsArray.length()];
            for(int i = 0; i < deviceEventsArray.length(); i++ ){
                if(deviceEventsArray.getInt(i) == 0){
                    devices[i] = Analytics.DeviceEvent.NONE;
                } else if(deviceEventsArray.getInt(i) == 1){
                    devices[i] = Analytics.DeviceEvent.ALL;
                } else if(deviceEventsArray.getInt(i) == 2){
                    devices[i] = Analytics.DeviceEvent.LIFECYCLE;
                } else if(deviceEventsArray.getInt(i) == 3){
                    devices[i] = Analytics.DeviceEvent.NETWORK;
                }
            }
            Analytics.init(app, applicationName, clientApiKey, hasContext, devices);
        } catch (JSONException e) {
            analyticsLogger.error("init :: Analytics failed to initialize. Please review arguments");
            callbackContext.error(e.getMessage());
        }

    }

    /**
     * Log an analytics event
     * @param args  JSONArray that contains the description for the event
     * @param callbackContext
     */

    public void log(final JSONArray args, final CallbackContext callbackContext){
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    JSONObject meta = args.getJSONObject(0);
                    Analytics.log(meta);
                } catch (JSONException e) {
                    analyticsLogger.error("log :: " + errorEmptyArg);
                    callbackContext.error(e.getMessage());
                }
            }
        });
    }

}

File: src/ios/CDVBMSAnalytics.swift
/*
*     Copyright 2016 IBM Corp.
*     Licensed under the Apache License, Version 2.0 (the "License");
*     you may not use this file except in compliance with the License.
*     You may obtain a copy of the License at
*     http://www.apache.org/licenses/LICENSE-2.0
*     Unless required by applicable law or agreed to in writing, software
*     distributed under the License is distributed on an "AS IS" BASIS,
*     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*     See the License for the specific language governing permissions and
*     limitations under the License.
*/


import Foundation
import BMSCore
import BMSAnalytics


@objc(CDVBMSAnalytics) class CDVBMSAnalytics : CDVPlugin {

    func setUserIdentity(_ command: CDVInvokedUrlCommand){
        let userIdentity = command.arguments[0] as! String

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                Analytics.userIdentity = userIdentity
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: true)
                // call success callback
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                Analytics.userIdentity = userIdentity
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool: true)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId: command.callbackId)
            })
        #endif
    }

    func enable(_ command: CDVInvokedUrlCommand) {

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                Analytics.isEnabled = true
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: true)
                // call success callback
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                Analytics.isEnabled = true
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool: true)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId: command.callbackId)
            })
        #endif
    }

    func disable(_ command: CDVInvokedUrlCommand) {

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                Analytics.isEnabled = false
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: false)
                // call success callback
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                Analytics.isEnabled = false
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool: false)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId: command.callbackId)
            })
        #endif
    }

    func isEnabled(_ command: CDVInvokedUrlCommand) {

        // has success, failure callbacks

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                let isEnabled = Analytics.isEnabled
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: isEnabled)
                // call success callback
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                Analytics.isEnabled = false
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool: false)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId: command.callbackId)
            })
        #endif
    }



    func send(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                Analytics.send(completionHandler: { (response: Response?, error:Error?) in
                        if (error != nil) {
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: false)
                            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                        } else {
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: true)
                            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                        }
                })
            })
        #else
            self.commandDelegate!.runInBackground({
                Analytics.send(completionHandler: { (response: Response?, error:NSError?) in
                    if (error != nil) {
                        // process the error
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsBool:false)
                        self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    } else {
                        // process success
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool: true)
                        self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    }
                })
            })
        #endif
    }

     func initialize(_ command: CDVInvokedUrlCommand){

            let appName = command.arguments[0] as! String
            let clientApiKey = command.arguments[1] as! String
            let hasUserContext = command.arguments[2] as! Bool
            let events = command.arguments[3] as! [Int]
            var deviceEvents = [DeviceEvent]()
            var lifecycleFlag: Bool = false
            var networkFlag:Bool = false
            var noneFlag:Bool = false

            for i in 0..<events.count {
                switch(events[i]){
                case 0:
                    noneFlag = true
                    break; // NONE should not enable any deviceEvents
                case 1:
                    lifecycleFlag = true;
                    networkFlag = true;
                    break;
                case 2:
                    lifecycleFlag = true
                    deviceEvents.append(.lifecycle)
                    break
                case 3:
                    networkFlag = true
                    deviceEvents.append(.network)
                    break
                default:
                    lifecycleFlag = true
                    deviceEvents.append(.lifecycle)
                    break
                }

            }

            #if swift(>=3.0)
                self.commandDelegate!.run(inBackground: {
                    if(noneFlag){
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext)
                    } else if (lifecycleFlag && networkFlag){
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext, deviceEvents: .lifecycle, .network)
                    } else if(networkFlag) {
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext, deviceEvents: .network)
                    } else if(lifecycleFlag){
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext, deviceEvents: .lifecycle)
                    }

                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:true)
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
            #else
                self.commandDelegate!.runInBackground({
                    if(noneFlag){
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext)
                    } else if (lifecycleFlag && networkFlag){
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext, deviceEvents: .lifecycle, .network)
                    } else if(lifecycleFlag){
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext, deviceEvents: .lifecycle)
                    }  else if(networkFlag) {
                        Analytics.initialize(appName: appName, apiKey: clientApiKey, hasUserContext: hasUserContext, deviceEvents: .network)
                    }

                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool:true)
                    self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                })
            #endif

     }


    func log(_ command: CDVInvokedUrlCommand) {
        let meta = command.arguments[0] as! Dictionary<String, Any>

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {
                Analytics.log(metadata: meta)
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:true)
                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({
                Analytics.log(metadata: (meta as? [String: AnyObject])!)
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsBool:true)
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif

    }
}
